src/btree/bt_sync.c

```diff
@@ -101,39 +98,114 @@ __sync_dup_walk(WT_SESSION_IMPL *session, WT_REF *walk, uint32_t flags, WT_REF *
 }
 
 /*
- * __sync_ref_is_obsolete --
- *     Return whether the ref is obsolete according to the newest stop time pair.
+ * __sync_dup_walk --
+ *     Duplicate a tree walk point.
+ */
+static inline int
+__sync_dup_walk(WT_SESSION_IMPL *session, WT_REF *walk, uint32_t flags, WT_REF **dupp)
+{
+    WT_REF *old;
+
+    if ((old = *dupp) != NULL) {
+        *dupp = NULL;
+        WT_RET(__wt_page_release(session, old, flags));
+    }
+
+    /* It is okay to duplicate a walk before it starts. */
+    if (walk == NULL || __wt_ref_is_root(walk)) {
+        *dupp = walk;
+        return (0);
+    }
+
+    return (__sync_dup_hazard_pointer(session, walk, dupp));
+}
+
+/*
+ * __sync_ref_list_add --
+ *     Add an obsolete history store ref to the list.
+ */
+static int
+__sync_ref_list_add(WT_SESSION_IMPL *session, WT_REF_LIST *rlp, WT_REF *ref)
+{
+    if (rlp->entry + 1 >= rlp->max_entry)
+        WT_RET(__wt_realloc_def(session, &rlp->max_entry, rlp->max_entry + 10, &rlp->list));
```

Comment:
There's no need to do the test before calling `__wt_realloc_def()`, see `__conn_foc_add()` for an example.

Reply:
Fixed.